### ISSUE TITLE
IndexOutofRangeException when read parquet file

### ISSUE THREAD

**Issue by guoliqiang** [aloneguid/parquet-dotnet#267]
Different parquet-dotnet versions have different behavior for my one parquet file as below (I have many parquet files; this only happen on one file)

No better what API I chose (e.g. class deserialize, row based, low API), the problem could be reproduced.

Version = 4.5.0 ~ 4.5.1
It will throw IndexOutofRangeException  exception

Version = 3.7.4
It will return wrong data, the long/int element value of the first row are all 1 (I am sure, the real value is not 1)

Version = 4.4.7
Everything looks good.

Code snippet using class deserialize API (As I mentioned, problem also could be reproduced with row based, low API)

 public class Sample
    {
        public int Id1 { get; set; }
        public long Id2 { get; set; }
        public long Id3 { get; set; }
        public long Id4 { get; set; }
        public long Id5 { get; set; }
        public long Id6 { get; set; }
        public short Id7 { get; set; }
        public Byte Id8 { get; set; }
        public short Id9 { get; set; }
        public long Id10 { get; set; }
        public Byte Id11 { get; set; }
        public DateTime myTime { get; set; }
        public Byte Id12 { get; set; }
        public string myString { get; set; }
    };

 using (FileStream fileStream = File.OpenRead(path))
 {
                var itemList = ParquetConvert.DeserializeAsync<Smaple>(Sample).Result;
                Console.WriteLine(itemList.Length);
 }

I am happy to share my parquet file offline to help the debug.  Thanks

By the way, I tried to debug using the latest version source code, here is the stack trace

This exception was originally thrown at this call stack:
    [External Code]
    Parquet.Encodings.BitPackedEncoder.Unpack8IntValuesLE1(System.Span<byte>, System.Span<int>) in BitPackedEncoder.cs
    Parquet.Encodings.BitPackedEncoder.Unpack8ValuesLE(System.Span<byte>, System.Span<int>, int) in BitPackedEncoder.cs
    Parquet.Encodings.BitPackedEncoder.Decode8ValuesLE(System.Span<byte>, System.Span<int>, int) in BitPackedEncoder.cs
    Parquet.Encodings.RleBitpackedHybridEncoder.DecodeBitpacked(int, System.Span<byte>, ref int, int, System.Span<int>, int) in RleBitpackedHybridEncoder.cs
    Parquet.Encodings.RleBitpackedHybridEncoder.Decode(System.Span<byte>, int, System.Span<int>, int) in RleBitpackedHybridEncoder.cs
    Parquet.Encodings.RleBitpackedHybridEncoder.Decode(System.Span<byte>, int, int?, out int, System.Span<int>, int) in RleBitpackedHybridEncoder.cs
    Parquet.File.DataColumnReader.ReadRleDictionary(System.Span<byte>, int, System.Span<int>) in DataColumnReader.cs
    Parquet.File.DataColumnReader.ReadColumn(System.Span<byte>, Parquet.Thrift.Encoding, long, int, Parquet.File.PackedColumn) in DataColumnReader.cs
    Parquet.File.DataColumnReader.ReadDataPageAsync(Parquet.Thrift.PageHeader, Parquet.File.PackedColumn, long) in DataColumnReader.cs
    ...
    [Call Stack Truncated]

**Comment by aloneguid** [comment]
Please send the sample to z0ks9srcl@mozmail.com

**Comment by edraltman** [comment]
I'm hitting the same **IndexOutofRangeException** when loading a particular column within a test file that worked in earlier versions of this package. The error is throwing from a different method.

### Test file
We use data from the nyc taxi dataset, which has been re-hosted by UrsaLabs on their public s3 bucket. I can reproduce this issue with their 2010-01 dataset file:

Blog page: https://mran.microsoft.com/snapshot/2020-04-15/web/packages/arrow/vignettes/dataset.html
S3 bucket: https://ursa-labs-taxi-data.s3.us-east-2.amazonaws.com/
dataset: https://ursa-labs-taxi-data.s3.us-east-2.amazonaws.com/2010/01/data.parquet

Loading this file will always fail on reading the **passenger_count** column, which is an Sbyte.
I created a test within ParquetReaderTest.cs in this package that attempts to read this file. Stacktrace:
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.

System.IndexOutOfRangeException
Index was outside the bounds of the array.
   at Parquet.Encodings.BitPackedEncoder.Unpack8IntValuesLE3(Span`1 src, Span`1 dest) in C:\dev\parquet-dotnet\src\Parquet\Encodings\BitPackedEncoder.cs:line 1258
   at Parquet.Encodings.BitPackedEncoder.Unpack8ValuesLE(Span`1 src, Span`1 dest, Int32 bitWidth) in C:\dev\parquet-dotnet\src\Parquet\Encodings\BitPackedEncoder.cs:line 161
   at Parquet.Encodings.BitPackedEncoder.Decode8ValuesLE(Span`1 src, Span`1 dest, Int32 bitWidth) in C:\dev\parquet-dotnet\src\Parquet\Encodings\BitPackedEncoder.cs:line 139
   at Parquet.Encodings.RleBitpackedHybridEncoder.DecodeBitpacked(Int32 header, Span`1 data, Int32& dataOffset, Int32 bitWidth, Span`1 dest, Int32 maxItems) in C:\dev\parquet-dotnet\src\Parquet\Encodings\RleBitpackedHybridEncoder.cs:line 213
   at Parquet.Encodings.RleBitpackedHybridEncoder.Decode(Span`1 data, Int32 bitWidth, Span`1 dest, Int32 pageSize) in C:\dev\parquet-dotnet\src\Parquet\Encodings\RleBitpackedHybridEncoder.cs:line 172
   at Parquet.Encodings.RleBitpackedHybridEncoder.Decode(Span`1 s, Int32 bitWidth, Nullable`1 spanLength, Int32& usedSpanLength, Span`1 dest, Int32 pageSize) in C:\dev\parquet-dotnet\src\Parquet\Encodings\RleBitpackedHybridEncoder.cs:line 142
   at Parquet.File.DataColumnReader.ReadRleDictionary(Span`1 s, Int32 maxReadCount, Span`1 dest) in C:\dev\parquet-dotnet\src\Parquet\File\DataColumnReader.cs:line 350
   at Parquet.File.DataColumnReader.ReadColumn(Span`1 src, Encoding encoding, Int64 totalValuesInChunk, Int32 totalValuesInPage, PackedColumn pc) in C:\dev\parquet-dotnet\src\Parquet\File\DataColumnReader.cs:line 285
   at Parquet.File.DataColumnReader.ReadDataPageAsync(PageHeader ph, PackedColumn pc, Int64 totalValuesInChunk) in C:\dev\parquet-dotnet\src\Parquet\File\DataColumnReader.cs:line 195
   at Parquet.File.DataColumnReader.ReadAsync(CancellationToken cancellationToken) in C:\dev\parquet-dotnet\src\Parquet\File\DataColumnReader.cs:line 63
   at Parquet.ParquetReader.ReadEntireRowGroupAsync(Int32 rowGroupIndex) in C:\dev\parquet-dotnet\src\Parquet\ParquetReader.cs:line 149
   at Parquet.Test.ParquetReaderTest.Read_Broken_Test_file(String parquetFile) in C:\dev\parquet-dotnet\src\Parquet.Test\ParquetReaderTest.cs:line 48
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 264
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
```

**Comment by aloneguid** [comment]
This is great, thanks.

**Comment by aloneguid** [comment]
v4.5.4 has a fix for this, please check ;)

**Comment by edraltman** [comment]
4.5.4 resolved this issue for me. Thanks for the rapid fix!

**Comment by guoliqiang** [comment]
> v4.5.4 has a fix for this, please check ;)

Thanks @aloneguid  v4.5.4 also works for me. 

Want to share more.  I found a workaround when using 3.7.4, although I do not know what happened exactly in the background.

1. all columns in my parquet files are nullable. (however, I am sure there is no null value in my parquet file)
2. thus I changed all Sample class elements definitions to nullable.

Above workaround works in 3.7.4
